src/gui/rss/automatedrssdownloader.cpp

```diff
@@ -452,13 +453,12 @@ void AutomatedRssDownloader::on_exportBtn_clicked()
             path += EXT_LEGACY;
     }
 
-    QFile file {path};
-    if (!file.open(QFile::WriteOnly)
-            || (file.write(RSS::AutoDownloader::instance()->exportRules(format)) == -1))
-            {
-        QMessageBox::critical(
-                    this, tr("I/O Error")
-                    , tr("Failed to create the destination file. Reason: %1").arg(file.errorString()));
+    const QByteArray rules {RSS::AutoDownloader::instance()->exportRules(format)};
+    QSaveFile file {path};
+    if (!file.open(QFile::WriteOnly) || (file.write(rules) != rules.length()) || !file.commit())
```

Comment:
IMO, it's time to provide some helper for this purpose:
```c++
void saveToFile(const QByteArray &data, const QString &path)
{
    QSaveFile file {path};
    if (!file.open(QSaveFile::WriteOnly) || (file.write(data) != data.size()) || !file.commit())
        throw RuntimeError(file.errorString());
}
```

Reply:
What about returning `std::optional<QString>` instead? I think using exceptions breaks the flow of code.

Reply:
PR updated.

Reply:
>I think using exceptions breaks the flow of code.

Yes. But isn't it a "feature" of exceptions?
Of course, it requires more care on the part of the programmer, but it provides more advanced features.

Otherwise you can use output parameter to provide additional error info (e.g. error message).

But I really dislike to `return` error message (unless you implement some kind of function that is really produces error message, e.g. `getErrorMessage(int errorCode)`).

Reply:
>Yes. But isn't it a "feature" of exceptions?
>Of course, it requires more care on the part of the programmer, but it provides more advanced features.

Exceptions provide a kind of handling which is viewed as "out of band". I wouldn't like it to be pervasive into functions, it should only be where it needs to be, here is not very well suited IMO.

>But I really dislike to return error message (unless you implement some kind of function that is really produces error message, e.g. getErrorMessage(int errorCode)).

Why? Other programming languages (Rust, Haskell) uses an `std::optional` like structure (`Result<T, E>`) for error handling. I think if the structure is good enough for them then it is good enough for C++.
Or maybe you are just uncomfortable with the idea of it (the structure or a type that carries some results...)?